File: Smart Sweepers/src/neuronalNet/NeuralNet.java
package neuronalNet;

import java.util.ArrayList;
import java.util.List;

import configuration.Parameters;

public class NeuralNet {
	private int numInputs;
	private int numOutputs;
	private int numHiddenLayers;
	private int neuronsPerHiddenLayer;

	List<SNeuronLayer> neuronLayers;

	public NeuralNet() {
		numInputs = Parameters.iNumInputs;
		numOutputs = Parameters.iNumOutputs;
		numHiddenLayers = Parameters.iNumHidden;
		neuronsPerHiddenLayer = Parameters.iNeuronsPerHiddenLayer;

		neuronLayers = new ArrayList<>();

		createNet();
	}

	public void createNet() {
		if (numHiddenLayers > 0) {

			// create first hidden layer
			neuronLayers.add(new SNeuronLayer(neuronsPerHiddenLayer, numInputs));

			for (int i = 0; i < numHiddenLayers - 1; i++) {
				neuronLayers.add(new SNeuronLayer(neuronsPerHiddenLayer, neuronsPerHiddenLayer));
			}

			// create output layer
			neuronLayers.add(new SNeuronLayer(numOutputs, neuronsPerHiddenLayer));
		} else {
			neuronLayers.add(new SNeuronLayer(numOutputs, numInputs));
		}
	}

	public List<Double> getWeights() {
		List<Double> weights = new ArrayList<>();

		for (int numLayer = 0; numLayer < neuronLayers.size(); numLayer++) {
			SNeuronLayer layer = neuronLayers.get(numLayer);
			for (int numNeuron = 0; numNeuron < layer.numNeurons; numNeuron++) {
				SNeuron neuron = layer.neurons.get(numNeuron);
				weights.addAll(neuron.inputWeight);
			}
		}
		return weights;
	}

	public void putWeights(List<Double> weights) {
		int cWeight = 0;

		for (int numLayer = 0; numLayer < neuronLayers.size(); numLayer++) {
			SNeuronLayer layer = neuronLayers.get(numLayer);
			for (int numNeuron = 0; numNeuron < layer.numNeurons; numNeuron++) {
				SNeuron neuron = layer.neurons.get(numNeuron);
				for (int numWeight = 0; numWeight < neuron.numInputs; numWeight++) {
					neuron.inputWeight.set(numWeight, weights.get(cWeight++));
				}
			}
		}
	}

	public int getNumberOfWeights() {
		int weights = 0;

		for (int numLayer = 0; numLayer < neuronLayers.size(); numLayer++) {
			SNeuronLayer layer = neuronLayers.get(numLayer);
			for (int numNeuron = 0; numNeuron < layer.numNeurons; numNeuron++) {
				SNeuron neuron = layer.neurons.get(numNeuron);
				weights += neuron.numInputs;
			}
		}
		return weights;
	}

	public List<Double> update(List<Double> inputs) {
		List<Double> outputs = new ArrayList<>();

		if (inputs.size() != numInputs) {
			return outputs;
		}

		for (int numLayer = 0; numLayer < neuronLayers.size(); numLayer++) {
			if (numLayer > 0) {
				inputs.clear();
				inputs.addAll(outputs);
			}

			outputs.clear();

			SNeuronLayer layer = neuronLayers.get(numLayer);
			for (int numNeuron = 0; numNeuron < layer.numNeurons; numNeuron++) {
				double netInput = 0;

				SNeuron neuron = layer.neurons.get(numNeuron);
				int numInputs = neuron.numInputs;

				for (int numWeight = 0; numWeight < numInputs - 1; numWeight++) {
					netInput += neuron.inputWeight.get(numWeight) * inputs.get(numWeight);
				}

				netInput += neuron.inputWeight.get(numInputs - 1) * Parameters.dBias;

				outputs.add(sigmoid(netInput, Parameters.dActivationResponse));

			}
		}

		return outputs;
	}

	private double sigmoid(double netInput, double dActivationResponse) {
		return (1 / (1 + Math.exp(-netInput / dActivationResponse)));
	}
}

File: Smart Sweepers/src/neuronalNet/NeuralNet.kt
package neuronalNet

import configuration.*
import generics.*
import utils.*
import kotlin.math.*

// we need an additional weight for the (neural-net) bias hence the +1
@Suppress("NOTHING_TO_INLINE")
private inline fun Size.bias(): Size = this + 1

private class Neuron(nominalInputSize: Size) {
    val inputSize: Size = nominalInputSize.bias()

    // initial values for weights are triangular random numbers (between -1 and 1) peaking at zero.
    val weights = MutableList(inputSize) { rand.randomClamped() }
}

private class Layer(layerSize: Size, inputsPerNeuron: Size) {
    val neurons = List(layerSize) { Neuron(inputsPerNeuron) }
}

private fun sigmoid(rawOutput: Double, dActivationResponse: Double) =
    1 / (1 + exp(-rawOutput / dActivationResponse))

class NeuralNet {
    companion object {
        // all the parameters below are determined lazily, so that they can be read into
        // Parameters before any NeuralNet instance is created.
        private val numInputs: Size by lazy {
            Parameters.iNumInputs.also { require(it > 0) { "inputs: $it, must be positive" } }
        }
        private val numOutputs: Size by lazy {
            Parameters.iNumOutputs.also { require(it > 0) { "outputs: $it, must be positive" } }
        }
        private val numHiddenLayers: Size by lazy {
            Parameters.iNumHidden.also { require(it >= 0) { "hidden layers: $it, must not be negative" } }
        }
        private val numNeuronsPerHiddenLayer: Size by lazy {
            Parameters.iNeuronsPerHiddenLayer.also { require(numHiddenLayers == 0 || it > 0) { "neurons per hidden layer: $it, must be positive when using hidden layers" } }
        }

        @JvmStatic
        val numberOfWeights by lazy {
            if (numHiddenLayers == 0) numOutputs * numInputs.bias()
            else numNeuronsPerHiddenLayer * numInputs.bias() +
                    (numHiddenLayers - 1) * numNeuronsPerHiddenLayer * numNeuronsPerHiddenLayer.bias() +
                    numOutputs * numNeuronsPerHiddenLayer.bias()
        }
    }

    private val layers = sequence {
        // with no hidden layers ...
        if (numHiddenLayers <= 0) {
            // ... there's only a single layer, so it has to directly convert to outputs from inputs
            yield(Layer(numOutputs, numInputs))
        } else {
            // with hidden layers, the first hidden (possibly only) must convert to the hidden from inputs ...
            yield(Layer(numNeuronsPerHiddenLayer, numInputs))
            // ... then any other hidden layers keep converting hidden from hidden ...
            repeat(numHiddenLayers - 1) { yield(Layer(numNeuronsPerHiddenLayer, numNeuronsPerHiddenLayer)) }
            // ... with the last layer converting to outputs from hidden
            yield(Layer(numOutputs, numNeuronsPerHiddenLayer))
        }
    }.toList()

    @Suppress("unused", "MemberVisibilityCanBePrivate")
    fun getWeights() = layers.flatMap { l -> l.neurons.flatMap { n -> n.weights } }

    fun putWeights(weightByAbsoluteIndex: ReadWeight) {
        var weightAbsoluteIndex = 0
        fun nextWeight() = weightByAbsoluteIndex(weightAbsoluteIndex)
            .also { weightAbsoluteIndex += 1 }

        layers.forEach { l ->
            l.neurons.forEach { n ->
                for (which in n.weights.indices) {
                    n.weights[which] = nextWeight()
                }
            }
        }
    }

    fun update(inputs: List<Double>) =
        layers.fold(inputs) { passed, l ->
            l.neurons.map { n ->
                n.weights.zip(passed.plus(Parameters.dBias))
                    .fold(0.0) { acc, weightInputPair -> acc + weightInputPair.first * weightInputPair.second }
                    .let { sigmoid(it, Parameters.dActivationResponse) }
            }
        }
}
